### ISSUE TITLE
TypeScript 3.6 Iteration Plan

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#31639]
This document outlines our focused tasks for TypeScript 3.6, as well as some of the discussion that explains how/why we prioritized certain work items. Nothing is set in stone, but we will strive to complete them in a reasonable timeframe.

Note that this release shifts TypeScript from a 2-month release cycle to a 3-month release cycle.

# Dates

* 3.5 ships (May 29th)
* TypeScript 3.6 Beta
    * Snap on July 12th
        * Kick off build for CTI
    * Release on July 16th
* First VS Code Insiders Build: August 1st?
* TypeScript 3.6 RC
    * Snap on August 9th
        * Kick off build for CTI
        * Kick off insiders for VS Code
    * Release on August 13th
* TypeScript 3.6 Final
    * Snap on August 23rd
        * Kick off build for CTI
        * Kick off insiders for VS Code
    * Release on August 27th 🚀

# Work Items

## Expected Work Items

* Bugs bugs bugs
* Language
    * [Strongly typed iterators and generators](https://github.com/Microsoft/TypeScript/issues/2983)
    * [Stricter types for `IteratorResult`](https://github.com/Microsoft/TypeScript/issues/11375)
    * [ECMAScript private instance fields](https://github.com/microsoft/TypeScript/pull/30829)
* Compiler
    * [APIs for `--incremental` builds and project references](https://github.com/Microsoft/TypeScript/issues/29978)
    * [More accurate array spread](https://github.com/microsoft/TypeScript/pull/31166)
    * [UX on `Promise`s](https://github.com/microsoft/TypeScript/issues/30646)
    * Investigate TypeScript plugin APIs
    * [`--declaration` and `--allowJs`](https://github.com/Microsoft/TypeScript/issues/7546)
* Infrastructure
    * Automated testing infrastructure for language service on real world code and DefinitelyTyped
    * Test tiering - faster local tests, longer CI runs
    * Release process overhaul
        * [Need to enable automated cherry-picking.](https://github.com/microsoft/TypeScript/issues/31590)
        * [Enable experimental feature branching](https://github.com/microsoft/TypeScript/issues/31591)
        * Must document the process here.
        * `typescript-bot` commenting that PRs don't have associated issues with a milestone
    * [Migrate repo to ESLint](https://github.com/Microsoft/TypeScript/issues/30553)
* Editor Features
    * Await feedback for more completion priority improvements
    * [Automatic Semicolon Deletion for auto-imports/refactors](https://github.com/microsoft/TypeScript/issues/19882)
    * [Auto-completion on `Promise`s](https://github.com/microsoft/TypeScript/issues/31450)
    * [Call hierarchy](https://github.com/microsoft/TypeScript/issues/31863)
    * [TSServer diagnostics command](https://github.com/microsoft/TypeScript/issues/31592)

## Deferred Work Items

*None from planning meeting*
______

# Planning Meeting Notes

## Motivations

* Bug backlog
* [Goals and current 6-month roadmap](https://github.com/Microsoft/TypeScript/issues/29288)
* GitHub user feedback (👍s)
* Feedback from customer interviews, social media, past iteration plans
* Visual Studio and Visual Studio Code feedback, as well as new functionality demands
* Actionable PRs (need to make a call)

## General (compiler/infrastructure/reliability)

* Bugs bugs bugs
    * We simply haven't caught up with these from 3.4/3.5.
    * Committed ✅
* [Strongly typed iterators and generators](https://github.com/Microsoft/TypeScript/issues/2983)
    * PR out ✅
* [Stricter types for `IteratorResult`](https://github.com/Microsoft/TypeScript/issues/11375)
    * PR out. ✅
* [ECMAScript private instance fields](https://github.com/microsoft/TypeScript/pull/30829)
    * PR out. ✅
* [APIs for composite projects (`--build` mode)](https://github.com/Microsoft/TypeScript/issues/29978)
    * PR out. ✅
* [More accurate array spread](https://github.com/microsoft/TypeScript/pull/31166).
    * PR out. ✅
* [UX on `Promise`s](https://github.com/microsoft/TypeScript/issues/30646)
    * High value for all users, useful to partner teams.
    * Committed ✅
* Investigate TypeScript plugin APIs 🏃
    * Ongoing.
* `--declaration` and `--allowJs`
    * Concerns with edge cases.
    * Can use the GitHub crawler to see how stable it is.
    * Do the work and let the crawler find the tail of bugs before we're ready to commit. ✅
* Testing infra
    * Automated testing infrastructure for language service on real world code and DefinitelyTyped
    * Test tiering - faster local tests, longer CI runs
    * Release process overhaul
        * Need to enable automated cherry-picking.
        * Enable experimental feature branching
        * Must document the process here.
        * `typescript-bot` commenting that PRs don't have associated issues with a milestone.
    * All of these, yes ✅
* [Migrate repo to ESLint](https://github.com/Microsoft/TypeScript/issues/30553)
    * ✅ Seems doable, @a-tarasyuk seems interested in helping!

## Productivity

* Editor Features
    * Await feedback for more completion priority improvements
    * [Automatic Semicolon Deletion for auto-imports/refactors](https://github.com/microsoft/TypeScript/issues/19882)
        * One of the most-demanded set of items
        * Committed ✅
    * [Auto-completion on `Promise`s](https://github.com/microsoft/TypeScript/issues/31450)
        * Seems useful for partner teams as well as JS users
        * Committed ✅
    * [Call hierarchy](https://github.com/Microsoft/vscode/issues/16110)
        * Ask from VS Code
        * Committed ✅
    * Diagnostics command
        * Seems like it would provide a low-cost high-value way for users to diagnose "TypeScript is slow in my editor."
        * Committed ✅

**Comment by fatcerberus** [comment]
What’s meant by “snap”?  I know it’s a football metaphor, but I’m not sure what it means in terms of release scheduling.

**Comment by weswigham** [comment]
"snap" short for "snapshot" as in "to capture (a release)" - nothing to do with football in this context, I think.

**Comment by fatcerberus** [comment]
Oh - I thought it was a football metaphor since the items immediately below are “kick off X”.  Thanks, snapshot makes more sense!

**Comment by weswigham** [comment]
“kick off X” as in "to start X" heh, fair - I think that turn of phrase actually does come from football.

**Comment by RyanCavanaugh** [comment]
TL;DR during the next over we're scrumming the uprights and chipping the wicket.

**Comment by DanielRosenwasser** [comment]
I thought it was about branches snapping like on a tree even though that actually makes no sense since the branches are growing, not breaking.

**Comment by falsandtru** [comment]
You should put up #32273 on the working list and address it.

**Comment by falsandtru** [comment]
You should release #32283 with the next release to make complete user experience of enhanced generators.

**Comment by ericanderson** [comment]
@DanielRosenwasser Is 3.6.3 ready to go out? Could we get an rc if not?

**Comment by DanielRosenwasser** [comment]
Would you be able to validate typescript@3.6.3-insiders.20190909 for us?

**Comment by ericanderson** [comment]
First pass of testing suggests this will solve our problem. I need to run the full suite of builds/tests over night (2x) to be sure but if it pans out, I'll push this insider build to our devs tomorrow.

**Comment by kjleitz** [comment]
@DanielRosenwasser The `typescript@3.6.3-insiders.20190909` release fixes our issues (https://github.com/vuejs/vue/issues/10455) as well (see also https://github.com/microsoft/TypeScript/issues/33164). Solved all of the 100+ build errors!

**Comment by DanielRosenwasser** [comment]
Thanks everyone! As a final update before I close this out:

- 3.6.3 is available.
- We may provide a 3.6.4 with changes related to tree transforms and generators.
- the 3.7 iteration plan is now available: https://github.com/microsoft/TypeScript/issues/33352
